doc/barclamp.md

```diff
@@ -0,0 +1,1224 @@
+  clean-slate deployment of your barclamp's controller side components.
+
+The best way to set up your scratch node is to make it a lonely node. To this
+end, you will need to re-run mkcloud after you have created your cloud:
+
+```
+nodenumberlonelynode=1 mkcloud setuplonelynodes crowbar_register
+
+```
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by your first commit's ID.
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack)
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+alias Cpatch="(cd $BARCLAMP_FORK_PATH; git show | patch -N -p1 --merge -d /opt/dell)"
+
+# Full patch: applies your fork completely against /opt/crowbar
+function Fpatch()
+  {
+  cd $BARCLAMP_FORK_PATH
+  git diff "$COMMIT^" HEAD | patch -N --merge -p1 -d /opt/dell
+  for f in /opt/dell/*.yml
+    do
+    ln -s $f /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    done
+  chmod 755 /opt/dell/bin/*
+  }
+
+sync_crowbar()
+  {
+  barclamp_install.rb $BARCLAMP_FORK_PATH &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ $BARCLAMP_NAME
```

Comment:
@aspiers: sure, will do!

@samirjorina: yes, there's a bit of a problem here indeed. Two problems actually:

The line you commented on causes the cookbook upload to fail completely. I don't think that would have caused the breakage you are observing, though.

That one is likely due to the second problem: I tried [Tom's suggestion](https://github.com/crowbar/crowbar/pull/2294#discussion_r93575431) a while ago and forgot to revert the `$BARCLAMP_FORK_PATH` in line 143  back to `/opt/dell/crowbar_framework/barclamps` before the next commit, effectively leaving the approach I tried there in place. Hence your problems with Cinder, Keystone and weird migrations are likely to be the situation described here: https://github.com/crowbar/crowbar/pull/2294#discussion_r94596298

Thank you for trying this and spotting the problem! I fixed both now.